### PR TITLE
fix 19120 - allow CT slicing via `alias this`

### PR DESCRIPTION
**Pull request by ghost**
While the feature is not very useful it is consistent with the fact that indexing works.

### CODE REVIEW

**Comment by UplinkCoder** [comment]
That needs a spec change.

File: src/dmd/expressionsem.d [hunk 1]
```diff
@@ -7589,6 +7589,13 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
         {
             if (exp.lwr || exp.upr)
             {
+                Type t = exp.e1.type.aliasthisOf();
```
Comment (maxhaton) [inline src/dmd/expressionsem.d:7592]: A comment indicating the purpose of the code is needed.

File: src/dmd/mtype.d [hunk 2]
```diff
@@ -2056,8 +2056,8 @@ extern (C++) abstract class Type : ASTNode
         if (s.isAliasDeclaration())
             s = s.toAlias();
 
-        if (s.isTupleDeclaration())
-            return null;
+        if (TupleDeclaration td = s.isTupleDeclaration())
+            return td.tupletype;
```
Comment (BorisCarvajal) [inline src/dmd/mtype.d:2060]: You should use td.getType() but still this won't work with tuples that have values (not only types).

**Comment by BorisCarvajal** [comment]
The case target seems weird, you are slicing the type of a struct (not a struct value) and getting a type. In this case `alias this` has no connection with the issue.

I think actually the real problem that you are trying to solve is already working:

```d
struct A(T...) {
    alias tup = T;
    alias tup this;
}

immutable a = A!(int, "asd", char, 123)();
pragma(msg, a[0]);
pragma(msg, a[1]);
pragma(msg, a[2..$]);
```
outputs:
```
int
asd
tuple((char), 123)
```

File: src/dmd/mtype.d (hunk 2 above)
Comment (BorisCarvajal) [reply src/dmd/mtype.d:2060]: And because the real problem in your test case has nothing to do with `alias this`, so this code is wrong.

**Comment by RazvanN7** [comment]
ping @TungstenHeart

**Comment by BorisCarvajal** [comment]
It seems the root issue is whether or not `opSlice/opIndex` (and maybe others) should work at static level, this really needs clarification in the spec.
I found a previous attempt with an old [comment](https://github.com/dlang/dmd/pull/3071#issuecomment-38593348) from Hara Kenji that well describes the situation.

> While the feature is not very useful it is consistent with the fact that indexing works.

Another inconsistency in the language, particularly this construct can conflict with the syntax that denotes an static array. But we have `static opCall` too, so in the end it's just a problem of spec/design determination.

**Pull request closed without merging**